File: 2015/go/day03/day03.go
package day03

type position struct {
	x int // x position
	y int // y position
}

func PresentDelivery(str string) int {

	var pos position = position{0, 0}
	var houses = []position{pos}
	for _, d := range str {
		switch d {
		case '>':
			pos.x += 1
		case '<':
			pos.x -= 1
		case '^':
			pos.y += 1
		case 'v':
			pos.y -= 1
		}

		if houseInList(pos, houses) == false {
			houses = append(houses, pos)
		}
	}
	return len(houses)
}

func houseInList(h position, list []position) bool {
	for _, p := range list {
		if p == h {
			return true
		}
	}
	return false
}

func RoboSanta(str string) int {

	var santa position = position{0, 0}
	var robo position = position{0, 0}
	var houses = []position{robo}
	for i, d := range str {
		if i%2 == 0 {
			switch d {
			case '>':
				santa.x += 1
			case '<':
				santa.x -= 1
			case '^':
				santa.y += 1
			case 'v':
				santa.y -= 1
			}

			if houseInList(santa, houses) == false {
				houses = append(houses, santa)
			}
		} else {
			switch d {
			case '>':
				robo.x += 1
			case '<':
				robo.x -= 1
			case '^':
				robo.y += 1
			case 'v':
				robo.y -= 1
			}

			if houseInList(robo, houses) == false {
				houses = append(houses, robo)
			}
		}
	}
	return len(houses)
}

File: 2015/julia/Day03.jl

using Base
using Test
using BenchmarkTools

function NextPosition(position, c)
    if c == '>'
        return position + 1
    elseif c == '<'
        return position - 1
    elseif c == 'v'
        return position - 1im
    elseif c  == '^'
        return position + 1im
    end
    return position
end

function Visits(instructions = readlines("day03.txt"))
    
    santa::Complex = 0+0im
    # robot::Complex = 0+0im

    houses = Vector{Complex}()
    append!(houses, santa)
    for (i,c) in enumerate(join(instructions))
        # if isodd(i)
        santa = NextPosition(santa, c)
        if (santa in houses) == false
            append!(houses, santa)
        end
        # else
        #     robot = NextPosition(robot, c)
        #     if robot in houses == false
        #         append!(houses, robot)
        #     end
        # end
    end
    return length(houses)
end

@test Visits(">") == 2
@test Visits("^>v<") == 4
@test Visits("^v^v^v^v^v") == 2
println(Visits())

function VisitsWithRobot(instructions = readlines("day03.txt"))
    
    santa::Complex = 0+0im
    robot::Complex = 0+0im

    houses = Vector{Complex}()
    append!(houses, santa)
    # append!(houses, robot)
    for (i,c) in enumerate(join(instructions))
        if isodd(i)
            santa = NextPosition(santa, c)
            if (santa in houses) == false
                append!(houses, santa)
            end
        else
            robot = NextPosition(robot, c)
            if (robot in houses) == false
                append!(houses, robot)
            end
        end
    end
    return length(houses)
end

@test VisitsWithRobot("^>") == 3
@test VisitsWithRobot("^>v<") == 3
@test VisitsWithRobot("^v^v^v^v^v") == 11
println(VisitsWithRobot())
